hist/hist/inc/THnChain.h

```diff
@@ -0,0 +1,116 @@
+// @(#)root/hist:$Id$
+// Author: Benjamin Bannier, August 2016
+
+/*************************************************************************
+ * Copyright (C) 1995-2016, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#ifndef ROOT_THnChain
+#define ROOT_THnChain
+
+#include <TObject.h>
+
+#include <string>
+#include <vector>
+
+class TAxis;
+class TH1;
+class TH2;
+class TH3;
+class THnBase;
+
+/** \class THnChain
+A class to chain together multiple histograms.
+
+This class allows to chain together any `THnBase`-derived (`THn` or `THnSparse`)
+histograms from multiple files. Operations on the axes and projections are
+supported. The intent is to allow convenient merging merging of projections
+of high-dimensional histograms.
+
+\code{.cpp}
+// `file1.root` and `file2.root` contain a `THnSparse` named `hsparse`.
+THnChain hs("hsparse");
+hs.AddFile("file1.root");
+hs.AddFile("file2.root");
+
+// Project out axis 0, integrate over other axes.
+TH1* h0 = hs.Projection(0);
+
+// Project out axis 0, integrate over other axes in their active ranges.
+hs.GetAxis(1)->SetRangeUser(0, 0.1); // select a subrange
+TH1* h0 = hs.Projection(0);
+\endcode
+*/
+
+class THnChain : public TObject
+{
+ public:
+   /// Default constructor.
+   ///
+   /// \param name name of the histogram to work on
+   THnChain(const char* name) : fName(name) {}
+
+   /// Add a new file to this chain.
+   ///
+   /// \param fileName path of the file to add
+   void AddFile(const char* fileName);
+
+   /// Get an axis from the histogram.
+   ///
+   /// \param i index of the axis to retrieve
+   ///
+   /// This function requires that a file containing the histogram was
+   /// already added with `AddFile`.
+   ///
+   /// Properties set on the axis returned by `GetAxis` are propagated to all
+   /// histograms in the chain, so it can e.g. be used to set ranges for
+   /// projections.
+   TAxis* GetAxis(Int_t i);
+
+   /// See `THnBase::Projection` for the intended behavior.
+   TH1* Projection(Int_t xDim, Option_t* option = "");
```

Comment:
`Projection()` should ideally be `const` *and* thread-safe. Do you *need* to call `SetupAxes()`?